pkg/agent/nodeportlocal/portcache/port_table_windows.go

```diff
@@ -0,0 +1,298 @@
+			// one with a valid port mapping.
+			klog.ErrorS(err, "Cannot bind to local port, skipping it", "port", nplPort.NodePort)
+			updateCloseState(protocols)
+			continue
+		}
+
+		npData := &NodePortData{
+			NodePort:  nplPort.NodePort,
+			PodPort:   nplPort.PodPort,
+			PodIP:     nplPort.PodIP,
+			Protocols: protocols,
+		}
+		for _, protocol := range nplPort.Protocols {
+			protocolSocketData := npData.FindProtocol(protocol)
+			if protocolSocketData == nil {
+				return fmt.Errorf("unknown protocol %s", protocol)
+			}
+			protocolSocketData.State = stateInUse
+		}
+		pt.NodePortTable[nplPort.NodePort] = npData
+		pt.PodEndpointTable[podIPPortFormat(nplPort.PodIP, nplPort.PodPort)] = pt.NodePortTable[nplPort.NodePort]
+	}
+	// retry mechanism as iptables-restore can fail if other components (in Antrea or other
+	// software) are accessing iptables.
+	go func() {
+		defer close(synced)
+		var backoffTime = 2 * time.Second
+		for {
+			if err := pt.syncRules(); err != nil {
+				klog.ErrorS(err, "Failed to restore iptables rules", "backoff", backoffTime)
+				time.Sleep(backoffTime)
+				continue
+			}
+			break
+		}
+	}()
+	return nil
+}
+
+func (d *NodePortData) UpdateCloseStates() error {
+	for idx := range d.Protocols {
+		protocolSocketData := &d.Protocols[idx]
+		switch protocolSocketData.State {
+		case stateClosed:
+			// already closed
+			continue
+		case stateInUse:
+			// should not happen
+			return fmt.Errorf("protocol %s is still in use, cannot update status to closed", protocolSocketData.Protocol)
+		case stateOpen:
+			protocolSocketData.State = stateClosed
+		default:
+			return fmt.Errorf("invalid protocol state")
+		}
+	}
+	return nil
+}
+
+func (pt *PortTable) DeleteRule(podIP string, podPort int, protocol string) error {
+	pt.tableLock.Lock()
+	defer pt.tableLock.Unlock()
+	data := pt.getEntryByPodIPPort(podIP, podPort)
+	if data == nil {
+		// Delete not required when the PortTable entry does not exist
+		return nil
+	}
+	numProtocolsInUse := 0
+	var protocolSocketData *ProtocolSocketData
+	for idx, pData := range data.Protocols {
+		if pData.State != stateInUse {
+			continue
+		}
+		numProtocolsInUse++
+		if pData.Protocol == protocol {
+			protocolSocketData = &data.Protocols[idx]
+		}
+	}
+	if protocolSocketData != nil {
+		if err := pt.PodPortRules.DeleteRule(data.NodePort, podIP, podPort, protocol); err != nil {
+			return err
+		}
+		protocolSocketData.State = stateOpen
+		numProtocolsInUse--
+	}
+	if numProtocolsInUse == 0 {
+		// Node port is not needed anymore: status update
+		if err := data.UpdateCloseStates(); err != nil {
+			return err
+		}
+		delete(pt.NodePortTable, data.NodePort)
+		delete(pt.PodEndpointTable, podIPPortFormat(podIP, podPort))
+	}
+	return nil
+}
+
+func (pt *PortTable) DeleteRulesForPod(podIP string) error {
+	pt.tableLock.Lock()
+	defer pt.tableLock.Unlock()
+	podEntries := pt.getDataForPodIP(podIP)
+	for _, podEntry := range podEntries {
+		for len(podEntry.Protocols) > 0 {
```

Comment:
Why not use for-loop with range podEntry.Protocols directly?

Reply:
@wenyingd  You can see the discussion here [https://github.com/antrea-io/antrea/pull/2448#discussion_r727485243](url)
If the second protocol deletion fails, we need to make sure the first one will not exist in the next round.